### PR TITLE
Development audio mode

### PR DESCRIPTION
**Pull request by zhemitu-amzn**
### Issue #, if available:
None

### Description of changes:
Disable NoAudio mode, as checked-in presence support will not go into the next GA.

#### Unit test coverage
* Class coverage: 86%
* Line coverage: 68%

#### Manual test cases (add more as needed):
* [x] Join meeting
* [x] Leave meeting
* [x] Rejoin meeting
* [x] Send audio
* [x] Receive audio
* [x] See active speaker indicator when speaking
* [x] Mute/Unmute self
* [x] See local mute indicator when muted
* [x] See remote mute indicator when other is muted
* [x] See audio video events
* [x] See signal strength changes
* [x] See media metrics received
* [x] See roster updates when remote attendees join / leave the meeting
* [x] Enable local video
* [x] See local video tile
* [x] See remote video tile
* [x] Switch camera
* [x] See remote screen sharing content with attendee name (Chrome tab)
* [x] Pause remote video tile
* [x] Resume remote video tile
* [x] First time audio permissions
* [x] First time video permissions

#### Screenshots, if available:

device-2021-11-30-130153

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: amazon-chime-sdk/src/main/java/com/amazonaws/services/chime/sdk/meetings/internal/audio/DefaultAudioClientController.kt [hunk 1]
```diff
@@ -148,21 +162,27 @@ class DefaultAudioClientController(
         val appInfo = AppInfoUtil.initializeAudioClientAppInfo(context)
 
         uiScope.launch {
+            muteMicAndSpeaker = false
```
Comment (ziyiz-amzn) [inline amazon-chime-sdk/src/main/java/com/amazonaws/services/chime/sdk/meetings/internal/audio/DefaultAudioClientController.kt]: `muteMicAndSpeaker` is initialized to false, and there is no place to set it to true. is this statement needed?

File: amazon-chime-sdk/src/test/java/com/amazonaws/services/chime/sdk/meetings/internal/AttendeeStatusTest.kt [hunk 2]
```diff
@@ -18,6 +18,6 @@ class AttendeeStatusTest {
 
     @Test
     fun `from should return null with invalid value from param`() {
-        Assert.assertNull(AttendeeStatus.from(4))
+        Assert.assertNull(AttendeeStatus.from(5))
```
Comment (ziyiz-amzn) [inline amazon-chime-sdk/src/test/java/com/amazonaws/services/chime/sdk/meetings/internal/AttendeeStatusTest.kt]: why change?

**Review by ziyiz-amzn — COMMENTED** [review]
Please make sure that the JS SDK change is in sync with mobile for the release.

**Comment by nggandhi** [comment]
Once the PR is approved, please keep a single commit that removes NoAudio support, so that it will be easier to revert that commit later when NoAudio support can be added back.

File: CHANGELOG.md [hunk 3]
```diff
@@ -2,10 +2,9 @@
 
 ### Added
 * Added APIs for Audio Video configuration i.e `AudioVideoConfiguration` to be used during a meeting session.
-* Added support for joining meetings without audio i.e `AudioMode.NoAudio` or with audio using one of `AudioMode.Mono16K`, `AudioMode.Mono48K` and `AudioMode.Stereo48K` audio modes.
+* Added support for joining meetings with audio using one of `AudioMode.Mono16K`, `AudioMode.Mono48K` and `AudioMode.Stereo48K` audio modes.
```
Comment (nggandhi) [inline CHANGELOG.md]: Maybe change it to "Added support for joining meetings using one of ...."?

File: CHANGELOG.md [hunk 4]
```diff
@@ -2,10 +2,9 @@
...
 * **Breaking** The `AudioMode.Stereo48K` will be set as the default audio mode if not explicitly specified when starting the audio session. Earlier, Mono/16KHz audio was the default and the only audio mode supported.
-* Added an optional method `onAttendeesJoinedWithoutAudio` in `RealtimeObserver` to communicate the status of attendees who joined without audio.
-* [Demo] Added ways to join a meeting without audio or with audio using various audio modes.
+* [Demo] Added ways to join a meeting with audio using various audio modes.
```
Comment (nggandhi) [inline CHANGELOG.md]: Same as above comment

File: amazon-chime-sdk/src/main/java/com/amazonaws/services/chime/sdk/meetings/internal/audio/DefaultAudioClientController.kt (hunk 1 above)
Comment (nggandhi) [reply amazon-chime-sdk/src/main/java/com/amazonaws/services/chime/sdk/meetings/internal/audio/DefaultAudioClientController.kt]: +1

File: app/src/main/java/com/amazonaws/services/chime/sdkdemo/fragment/MeetingFragment.kt [hunk 5]
```diff
@@ -268,7 +268,7 @@ class MeetingFragment : Fragment(),
         } ?: AudioMode.Stereo48K
         val audioVideoConfig = AudioVideoConfiguration(audioMode = audioMode)
         // Update the Mic & Speaker states
-        updateLocalAttendeeAudioState(audioEnabled = audioVideoConfig.audioMode != AudioMode.NoAudio)
+        updateLocalAttendeeAudioState(audioEnabled = true)
```
Comment (nggandhi) [inline app/src/main/java/com/amazonaws/services/chime/sdkdemo/fragment/MeetingFragment.kt]: We may remove this method if we don't want to give any hint about NoAudio support

File: amazon-chime-sdk/src/main/java/com/amazonaws/services/chime/sdk/meetings/audiovideo/audio/AudioMode.kt [hunk 6]
```diff
@@ -0,0 +1,31 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package com.amazonaws.services.chime.sdk.meetings.audiovideo.audio
+
+/**
+ * [AudioMode] describes the audio mode in which the audio client should operate during a meeting session.
```
Comment (achronop) [inline amazon-chime-sdk/src/main/java/com/amazonaws/services/chime/sdk/meetings/audiovideo/audio/AudioMode.kt:9]: We need to be more specific. We need to mention that it affect the sample rate and the channel count fo the output (speaker) and the sample rate of the input (michrophone). Maybe we need to mention the sample rate and channel count values for input and output in every value aka Mono16K, Mono48K etc.

File: amazon-chime-sdk/src/main/java/com/amazonaws/services/chime/sdk/meetings/audiovideo/audio/AudioMode.kt [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+ */
+enum class AudioMode(val value: Int) {
```
Comment (achronop) [inline amazon-chime-sdk/src/main/java/com/amazonaws/services/chime/sdk/meetings/audiovideo/audio/AudioMode.kt:11]: I am wondering why we do not support Stereo16K. It should work.

File: amazon-chime-sdk/src/main/java/com/amazonaws/services/chime/sdk/meetings/internal/audio/DefaultAudioClientController.kt [hunk 8]
```diff
@@ -49,23 +51,34 @@ class DefaultAudioClientController(
...
+        }
         audioClient.sendMessage(
             AudioClient.MESS_SET_IO_SAMPLE_RATE,
-            AudioClient.AUDIO_CLIENT_SAMPLE_RATE
+            samplingRateConfig
         )
 
         // Result is in bytes, so we divide by 2 (16-bit samples)
+        // Result is in bytes, so we divide by 2 (16-bit samples)
+        val channelConfig = when (audioMode) {
```
Comment (achronop) [inline amazon-chime-sdk/src/main/java/com/amazonaws/services/chime/sdk/meetings/internal/audio/DefaultAudioClientController.kt]: `outputChannelConfig` would be more descriptive.

**Review by achronop — COMMENTED** [review]
Looks good, a few questions/suggetions.

File: amazon-chime-sdk/src/main/java/com/amazonaws/services/chime/sdk/meetings/internal/audio/DefaultAudioClientController.kt (hunk 1 above)
Comment (zhemitu-amzn) [reply amazon-chime-sdk/src/main/java/com/amazonaws/services/chime/sdk/meetings/internal/audio/DefaultAudioClientController.kt]: Removed

File: amazon-chime-sdk/src/test/java/com/amazonaws/services/chime/sdk/meetings/internal/AttendeeStatusTest.kt (hunk 2 above)
Comment (zhemitu-amzn) [reply amazon-chime-sdk/src/test/java/com/amazonaws/services/chime/sdk/meetings/internal/AttendeeStatusTest.kt]: Fixed. Changed back to 4.

File: CHANGELOG.md (hunk 3 above)
Comment (zhemitu-amzn) [reply CHANGELOG.md]: Fixed

File: CHANGELOG.md (hunk 4 above)
Comment (zhemitu-amzn) [reply CHANGELOG.md]: Fixed

File: app/src/main/java/com/amazonaws/services/chime/sdkdemo/fragment/MeetingFragment.kt (hunk 5 above)
Comment (zhemitu-amzn) [reply app/src/main/java/com/amazonaws/services/chime/sdkdemo/fragment/MeetingFragment.kt]: Removed

File: amazon-chime-sdk/src/main/java/com/amazonaws/services/chime/sdk/meetings/audiovideo/audio/AudioMode.kt (hunk 6 above)
Comment (zhemitu-amzn) [reply amazon-chime-sdk/src/main/java/com/amazonaws/services/chime/sdk/meetings/audiovideo/audio/AudioMode.kt:9]: Added more descriptions about audio mode.

File: amazon-chime-sdk/src/main/java/com/amazonaws/services/chime/sdk/meetings/audiovideo/audio/AudioMode.kt (hunk 7 above)
Comment (zhemitu-amzn) [reply amazon-chime-sdk/src/main/java/com/amazonaws/services/chime/sdk/meetings/audiovideo/audio/AudioMode.kt:11]: There was a discussion a while ago. The main concern is that stereo at 16kHz sampling rate does not have good quality, and could cause confusion to builders.

File: amazon-chime-sdk/src/main/java/com/amazonaws/services/chime/sdk/meetings/internal/audio/DefaultAudioClientController.kt (hunk 8 above)
Comment (zhemitu-amzn) [reply amazon-chime-sdk/src/main/java/com/amazonaws/services/chime/sdk/meetings/internal/audio/DefaultAudioClientController.kt]: Modified to use this name.

File: amazon-chime-sdk/src/main/java/com/amazonaws/services/chime/sdk/meetings/internal/audio/DefaultAudioClientController.kt [hunk 9]
```diff
@@ -49,23 +51,34 @@ class DefaultAudioClientController(
...
+            AudioMode.Stereo48K -> 48000
+        }
         audioClient.sendMessage(
             AudioClient.MESS_SET_IO_SAMPLE_RATE,
-            AudioClient.AUDIO_CLIENT_SAMPLE_RATE
+            samplingRateConfig
         )
 
         // Result is in bytes, so we divide by 2 (16-bit samples)
+        // Result is in bytes, so we divide by 2 (16-bit samples)
```
Comment (ziyiz-amzn) [inline amazon-chime-sdk/src/main/java/com/amazonaws/services/chime/sdk/meetings/internal/audio/DefaultAudioClientController.kt]: duplicate comments?

**Review by ziyiz-amzn — APPROVED** [review]
Looks good to me. Left on minor comment.

Note that you will need to run our integration automation test once you have a private build following this https://github.com/awslabs/amazon-chime-sdk-apple/tree/automated-test.
Otherwise, let the mobile sdk team know once you have a private build. (for both iOS and Android)

File: amazon-chime-sdk/src/main/java/com/amazonaws/services/chime/sdk/meetings/internal/audio/DefaultAudioClientController.kt (hunk 9 above)
Comment (zhemitu-amzn) [reply amazon-chime-sdk/src/main/java/com/amazonaws/services/chime/sdk/meetings/internal/audio/DefaultAudioClientController.kt]: Yes. Will get rid of the duplicate before merging.

**Review by achronop — APPROVED** [review]

**Review by nggandhi — APPROVED** [review]

**Pull request merged**